guide: DVC Experiments Overview

jorgeorpinel opened the pull request:
> Originally extracted from https://github.com/iterative/dvc.org/pull/2654#issuecomment-936772206

- The [guide's index](https://dvc-org-exp-dvc-exps-pa-qhfdse.herokuapp.com/doc/user-guide/experiment-management) goes over Exp Mgmt in general (not only `exp` commands).
- The [new Exps Overview page](https://github.com/iterative/dvc.org/pull/2909/files#diff-59c6f50a088ff9d6ffbb2a0cb284901ea7c6c732fda9b937dbb34df43a8b883c) explains the basic  `dvc exp` workflow (happy path), contains implementation details (extracted [from Running Exps](https://github.com/iterative/dvc.org/pull/2909/files#diff-213c105f41cfb7e3abe29cba4fd80d9cf459bb94703b0a49708b23cf6b6cf5c5)), and **introduces `exp init`**.
- Additionally, [Persisting Exps](https://github.com/iterative/dvc.org/pull/2909/files#diff-64edd45172545bed656fae31b6639b4f9cf6fe51853eeca8f9208eb161527bbd) is now in a single place (stand-alone guide).

OTHER possible goals (may follow-up after this PR if needed):
- [x] Explain about exp ~~IDs and~~ names? From #2768
- [ ] What to say about **Organization patterns**?
  See https://github.com/iterative/dvc.org/pull/2909#pullrequestreview-775621414
  and https://github.com/iterative/dvc.org/pull/2909#pullrequestreview-812329316
- [ ] Improve **Basic workflow** section motivation and clearly differentiate it from the "new" DVC Experiments workflow.
  See https://github.com/iterative/dvc.org/pull/2909#pullrequestreview-819618983
- [ ] Create a diagram for different `exp` workflows? (at least the basic one)
  See https://github.com/iterative/dvc.org/pull/2909#pullrequestreview-794880808 below.
- [ ] Longer `exp init` section? Even rename or split from Overview? See [possible structure](https://github.com/iterative/dvc.org/pull/3015#pullrequestreview-804151331).

jorgeorpinel commented on content/docs/user-guide/experiment-management/dvc-experiments.md:
```diff
@@ -0,0 +1,36 @@
+## DVC Experiments
+
+_New in DVC 2.0_
```
> I think it's fine to use this (similar to DVC Remotes, etc)

OK so it seems we agree to keep this title and the branded term, in general.

> When we talk about the UG structure I have mixed feelings.
`Experiment Management -> DVC Experiments`
should be something like Overview?

You're right about this @shcheklein. Going for **DVC Experiments Overview** for now.

jorgeorpinel replied:
> need to think about better ideas.
Is the intention of this page is to give some overview btw?

The intention was dual originally:

1. Use the index to explain general experimentation strategies (currently run one by one, queue/parallel, checkpoints/live metrics; regular git workflow i.e. "persistent" exps) which ties up with Organization Patterns (automatic custom refs vs. tags/branches vs. directories [vs. multiple repos?](https://dvc.org/doc/user-guide/experiment-management/sharing-experiments#example-creating-a-directory-for-an-experiment) vs. hybrid)

2. Have a separate page (currently "Overview") to explain the main workflow (happy path): run, share, compare, persist? + clean-up (each one has it's own page now) based mostly on `exp` sub-commands. We can also use this page to put implementation details about experiments (Git custom refs, etc.)

I need to reiterate on this now that we have more guides. ⌛

jorgeorpinel replied:
OK I rewrote each page, PTAL @shcheklein 

Note that I also had to dissolve the existing Persisting Exps section into the stand-alone guide (and update a bunch of links).

jorgeorpinel replied:
p.s. please see the update PR description for links to easily review the main changes (last 4 [changed files](https://github.com/iterative/dvc.org/pull/2909/files)).

jorgeorpinel commented on content/docs/user-guide/experiment-management/index.md:
```diff
@@ -1,89 +1,72 @@
+If it never happened before, the stage command(s) are executed normally. Every
+subsequent time a [stage](/doc/command-reference/run) runs under the same
+conditions, the previous results can be restored instantly, without wasting time
+or computing resources.
```
> why do we add technical details to the index page then?

In this case I believe you're referring to the run cache? That's not part of DVC Experiments so we don't want run cache details in the dedicated Experiments Overview page I think.

But indeed I can try to make this more generic for now and leave all run-cache details in https://dvc.org/doc/user-guide/project-structure/internal-files#run-cache ⌛

jorgeorpinel replied:
Upon taking a closer look there's only a small detail here (run cache dir location) and everything else is in it's own guide section. I just made that existing section into an H2 so it has a ToC entry for now.

jorgeorpinel commented on content/docs/user-guide/project-structure/internal-files.md:
```diff
@@ -134,7 +134,7 @@ $ cat .dvc/cache/6f/db5336fce0dbfd669f83065f107551.dir
 That's how DVC knows that the other two cached files belong in the directory.
 
-### Run-cache
+## Run-cache
```
Note that this is linked to from a tooltip in https://dvc-org-exp-dvc-exps-pa-qhfdse.herokuapp.com/doc/user-guide/experiment-management#run-cache-automatic-log-of-stage-runs.

shcheklein commented on content/docs/command-reference/exp/apply.md:
```diff
@@ -30,9 +29,11 @@ results in the workspace.
+Note that the history of [checkpoints] found in the `experiment` is **not**
+preserved when applying and committing it.
+
+[persistent]: /doc/user-guide/experiment-management/persisting-experiments
```
unrelated changes?

shcheklein commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,27 @@
+# DVC Experiments Overview
```
it can be a good page (still not clear if need a separate one for this though, considering that we have index)

should we do a diagram here with the basic workflow?

should we include cleaning here?

jorgeorpinel replied:
> a diagram for the workflow would be nice

Since finalizing that properly may involve including alternative paths (a sort of flow chart) and design work, I vote to make it a follow-up issue (tied to the Exp Versioning release cc @dberenbaum) so we can merge this, if the content is approved in general.

shcheklein commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,27 @@
+
+[run them]: /doc/user-guide/experiment-management/running-experiments
+
+## Basic Workflow
```
Should we name the whole section as "Basic workflow" if that's what this section is about anyway

jorgeorpinel replied:
> name the whole section as "Basic workflow" if that's what this section is about

There is another intention here actually, which comes before the workflow. From https://github.com/iterative/dvc.org/pull/2909#discussion_r740633633:

>2... document implementation details about experiments (Git custom refs, etc.).

I just didn't do it yet because I thought it's planned in https://github.com/iterative/dvc.org/issues/2768 and didn't want to further complicate this PR which again is starting to grow beyond the original scope in terms of file changes... Maybe I'll start a nested one just to show what it looks like ⌛

jorgeorpinel replied:
> > document implementation details about experiments
>
> I'll start a nested one just to show what it looks like

@shcheklein @iesahin please see https://github.com/iterative/dvc.org/pull/3006/files for this purpose. If you approve both PRs I'll merge the nested one first and then this one.

shcheklein commented on content/docs/user-guide/experiment-management/index.md:
```diff
@@ -1,113 +1,82 @@
+> introduction to DVC experiments.
 
-### Organization patterns
+### Organization Patterns
```
why this change?

jorgeorpinel replied:
Mistake. Fixing:
```suggestion
### Organization patterns
```

iesahin commented on content/docs/user-guide/experiment-management/index.md:
```diff
@@ -1,113 +1,82 @@
-  cleans the slate so you can repeat the process.
-
-## Checkpoints in source code
+_New in DVC 2.0_
```
Is there anyone still using DVC 1.x? :)

jorgeorpinel replied:
Probably. We'll keep these for a year after the 2.0 release at least I think.

jorgeorpinel commented:
OK merged https://github.com/iterative/dvc.org/pull/3006#issuecomment-962537013 @shcheklein

shcheklein commented on content/docs/user-guide/experiment-management/persisting-experiments.md:
```diff
@@ -1,11 +1,9 @@
-
-In this section, we describe how to bring them to the standard Git workflow with
-`dvc exp branch` and `dvc exp apply`.
+DVC Experiments run outside of the regular Git workflow for faster iteration and
```
if we start with DVC experiments as first entry point to experiments people won't know about "Git" commit based experiments. Phrases like `run outside of the regular Git workflow` to some extent are fine for the former DVC users but will be confusing for new users.

jorgeorpinel replied:
> if we start with DVC experiments as first entry point to experiments

[This document](https://dvc-org-exp-dvc-exps-pa-qhfdse.herokuapp.com/doc/user-guide/experiment-management/persisting-experiments) is not a first entry point to the topic of experiments though. That's why there's an [index page](https://dvc-org-exp-dvc-exps-pa-qhfdse.herokuapp.com/doc/user-guide/experiment-management) and an [overview page](https://dvc-org-exp-dvc-exps-pa-qhfdse.herokuapp.com/doc/user-guide/experiment-management/experiments-overview) (WIP).

jorgeorpinel replied:
> Phrases like "run outside of the regular Git workflow" ... will be confusing for new users

This could be more emphasized in the overview page though, so it makes sense here (updated in 73175a99 for now).

iesahin commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,46 @@
+
+</details>
+
+Experiments will have an auto-generated ID like `exp-bfe64` by default. A custom
```
`auto-generated ID` is actually `auto-generated name`. We discussed this in another PR, considering there are too many different things that can be called ID, I think it might be easier for the reader just to call these _name_.

jorgeorpinel replied:
Agreed. Updated @iesahin

iesahin commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,46 @@
+goes like this:
+
+- Modify <abbr>parameters</abbr> or other dependencies (input data, source code,
+  stage definitions, etc.) of committed stages.
```
Modifying parameters and running the experiments occur at the same time with  `--set-param` option.

jorgeorpinel replied:
That's fine, this is a simplified and generalized description of the "basic workflow" for dvc exps.

iesahin commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,46 @@
+- [Run experiments] with `dvc exp run` (instead of `repro`) to execute the
+  pipeline. The results are reflected in your <abbr>workspace</abbr>, and
+  tracked automatically.
+- Use [metrics](/doc/command-reference/metrics) to identify the best
```
`dvc exp show` is used to identify the best experiments. Metrics provide a limited functionality compared to `exp show`.

jorgeorpinel replied:
`exp show` displays metrics.

jorgeorpinel replied:
I changed the link to /doc/user-guide/experiment-management/comparing-experiments now. I think that should resolve your concern.

jorgeorpinel replied:
Never mind, rolling that back since we specifically mention `exp show` in the next bullet...

jorgeorpinel replied:
Going back to this @iesahin , I merged the 2 bullets so the concern should be resolved now, I believe.

iesahin commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,46 @@
+  tracked automatically.
+- Use [metrics](/doc/command-reference/metrics) to identify the best
+  experiment(s).
+- Visualize and compare experiments with `dvc exp show` or `dvc exp diff`.
```
I think we shouldn't call _visualize_ for the work performed by `exp show` and `exp diff`. `dvc plots`, or the DVCLive HTML report can be used for visualization.

jorgeorpinel replied:
Changed to "review".

iesahin commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,46 @@
+- Visualize and compare experiments with `dvc exp show` or `dvc exp diff`.
+  Repeat 🔄
+- Make certain experiments [persistent] by committing their results to Git. This
+  cleans the slate so you can repeat the process later.
```
I think _cleaning the slate_ is a bit misleading, the experiments stay until we make Git branches from them, or delete them. Committing some experiments to Git doesn't delete the others.

jorgeorpinel replied:
Updated.

iesahin commented on content/docs/user-guide/experiment-management/index.md:
```diff
@@ -1,113 +1,82 @@
 [Parameters](/doc/command-reference/params) are simple values you can tweak in a
-human-readable text file, which cause different behaviors in your code and
-models. On the other end, [metrics](/doc/command-reference/metrics) (and
+formatted text file; They cause different behaviors in your code and models. On
```
"cause different behaviors" -> "may modify the results" or something softer?

jorgeorpinel replied:
Changing to "simple values in a formatted text file which you can tweak and use in your code". WDYT @iesahin ?

iesahin commented on content/docs/user-guide/experiment-management/index.md:
```diff
@@ -1,113 +1,82 @@
-human-readable text file, which cause different behaviors in your code and
-models. On the other end, [metrics](/doc/command-reference/metrics) (and
+formatted text file; They cause different behaviors in your code and models. On
+the other end, [metrics](/doc/command-reference/metrics) (and
```
I believe there is no need to mention metrics in experiments documentation. They are mostly superseded by `dvc exp show` and `diff`.

jorgeorpinel replied:
Here we are just mentioning the concept of metrics. It links to the cmd ref because we don't have a guide, basic concept page (yet), or even a tooltip for `<abbr>`. Ideally it should be one of those but it's out of scope for this PR.

iesahin replied:
Ok. I believe linking to the ref defeats the purpose of "mentioning the concept of metrics" but it's a matter of preference.

jorgeorpinel replied:
Not sure I follow. How can you link if you don't mention?

iesahin commented on content/docs/user-guide/experiment-management/index.md:
```diff
@@ -1,113 +1,82 @@
+It's up to you to decide how to organize completed experiments. These are the
+main alternatives:
 
 - **Git tags and branches** - use the repo's "time dimension" to distribute your
```
I think `dvc exp` commands work towards this organization pattern, not separate directories. IMO we can modify this section to describe the organization pattern DVC leads. We don't have much facility to use "space dimension" for experiments.

jorgeorpinel replied:
This index tries to cover the traditional experiments as well. It's not exclusively about `dvc exp` until you get to the sub-pages. That's one of the reasons for creating a separate DVC Experiments overview page in this PR.

jorgeorpinel replied:
It would be great to improve on this indeed so it's clear that we're separating manual exp tracking you can do on DVC projects vs. the DVC Experiments workflow. Similar to what's described in https://www.notion.so/iterative/Experiment-Workflows-3873c4f3cc2e49c6a7871a831bc8302b ⌛

jorgeorpinel replied:
> When I look into the text, I think DVC helps to organize the experiments in "space dimension" as well.

It can, for example multiple via `dvc.yaml` files (copy/pasted + small changes) or via `init --subrepo` (monorepo structure). Both supported by DVC

jorgeorpinel replied:
p.s. this makes me thing about another route: using `foreach` stages to quickly define multiple experiments based on a params file and running them all (in parallel). I guess it's a pre-`exp` way to manage experiments with DVC, but in which you can see all the results at once in your workspace (may be messy unless you create a bunch of subdirectories so perhaps it's the same as the  "space dimension")...

jorgeorpinel replied:
p.p.s I added a section about custom labels as well (for org pattern) based on [this table](https://docs.google.com/presentation/d/1C_owNoC72GvrpyMGlonHEYJ9I2rl2SLHkZQDMx0eT7A/edit#slide=id.gcb78e52e40_0_635). See c68fc784

iesahin replied:
> It can @iesahin, for example multiple via dvc.yaml files (copy/pasted + small changes) or via init --subrepo (monorepo structure). Both supported by DVC

The `dvc exp` workflow is orthogonal to these features, they are not alternatives to each other. One can use multiple `dvc.yaml` files with `dvc exp` as well.

jorgeorpinel replied:
They are conceptual alternatives which you can combine, which is already stated in the text.

jorgeorpinel commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,67 @@
+[compare]: /doc/user-guide/experiment-management/comparing-experiments
+[persistent]: /doc/user-guide/experiment-management/persisting-experiments
+
+## Initialize DVC Experiments on any project
+
+DVC Experiments features build on basic semantics of <abbr>DVC projects</abbr>.
+This means that minimal formalities are required, such as codifying a pipeline
+with `dvc.yaml` (even if it has a single <abbr>stage</abbr> that represents your
```
OK guys as agreed I've incorporated a section about `exp init` in this page as well. @dberenbaum if you could review [this one page](content/docs/user-guide/experiment-management/experiments-overview.md) that's be great.

> Unfortunately we ended with a large and mixed PR again, and now it's also tied to the Experiment Versioning release due to this change. So it'd be great to try and finalize it soon but if I need to split it once again I can. Lmk if any of you think so.

dberenbaum replied:
Wow, quick work getting this into the guide!

jorgeorpinel replied:
Thanks! It's just a couple paragraphs 😬. It may need more detail but I'm not super familiar with the latest iteration of the feature ATM.

dberenbaum commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,67 @@
+### ⚙️ How does DVC track experiments?
+
+Experiments are custom [Git references] (found in `.git/refs/exps`) with a
+single commit based on `HEAD` (not checked out by DVC). Note that these commits
```
Experiments with checkpoints have multiple commits.

> (not checked out by DVC)

What's not checked out by DVC? `HEAD`? The experiment?

dberenbaum replied:
It might be good to mention the motivation for using custom refs and some basic analogy or description of what they are. There are some good examples in [#2979#discussion_r757848241](https://github.com/iterative/dvc.org/pull/2979#discussion_r757848241).

jorgeorpinel replied:
Clarified.

jorgeorpinel replied:
Not sure we need motivation for implementation details.

jorgeorpinel replied:
p.s. I expanded on the general motivation a little based on the link. See d2c01119.

dberenbaum commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,67 @@
+</details>
+
+[run them]: /doc/user-guide/experiment-management/running-experiments
+[git references]: https://git-scm.com/book/en/v2/Git-Internals-Git-References
```
Should we link to Peter's blog post?

jorgeorpinel replied:
Great idea! Done.

dberenbaum commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,67 @@
+
+## Basic workflow
+
+`dvc exp` commands let you automatically track a variation to a committed
+project version (baseline). You can create independent groups of experiments
```
```suggestion
`dvc exp` commands let you automatically track a variation to a git commit.
You can create independent groups of experiments
```

Minor suggestion to make this a little more direct. I'm not sure it's clear what a "project version" is or what "baseline" means here.

jorgeorpinel replied:
Yeah it's more direct but it focuses too much on Git. It's not about Git commits.

"baseline" is defined in the previous paragraph.

I just removed "committed" instead (for now). WDYT?

dberenbaum commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,67 @@
+This means that minimal formalities are required, such as codifying a pipeline
+with `dvc.yaml` (even if it has a single <abbr>stage</abbr> that represents your
+entire process). Another typical preparation step is to create or modify a
+structured <abbr>parameters</abbr> file.
```
Should we also mention metrics? I think parameters and metrics are most important, since they may require writing additional code to read/write files, and it might be good to address this directly to avoid confusion.

jorgeorpinel replied:
Sure, added. PTAL

dberenbaum commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,67 @@
+entire process). Another typical preparation step is to create or modify a
+structured <abbr>parameters</abbr> file.
+
+`dvc exp init` lets you onboard any existing data science project to use DVC
+Experiments without having to worry bootstrapping DVC manually. It will prompt
+you a few simple questions and create a simple `dvc.yaml` file as well as other
+<abbr>metafiles</abbr> with sane default values. You can review these and commit
+them to Git to begin using DVC Experiments.
```
Consider moving this paragraph before the other one. Introduce the command to get started before explaining the requirements of DVC projects.

jorgeorpinel replied:
Yeah I guess that's more direct. Done (required a few adjustments), PTAL 2

dberenbaum reviewed (commented):
Awesome job incorporating `exp init` already, and the overall structure looks good!

dberenbaum commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,67 @@
+goes like this:
+
+- Modify hyperparameters or other dependencies (input data, source code,
+  commands to execute, etc.). Leave these changes un-committed in Git.
```
It might help to discuss why we have this workflow and want to leave changes un-committed. In other experiment tracking tools, the workflow looks like:

1. Run the experiment.
2. Log the commit hash of the current HEAD, even though it doesn't include your dirty changes.
3. If everything is successful, save changes in a new commit.

This creates a confusing state where the experiment should really be associated with the second commit instead of the first.

It might be too much detail or inappropriate for the page, but maybe it can be summarized, or it might spark other ideas. This is probably a good point for the blog post but not sure if there's a place for it...

jorgeorpinel replied:
> Similar to the discussion up in https://github.com/iterative/dvc.org/pull/2909#pullrequestreview-812329316

Yeah this needs more work. I'm hoping for now it's mergeable but I'll improve is as much as possible ⌛

jorgeorpinel replied:
> good point for the blog post

Do you mean linking from here? It's probably more important to link from the blog post to our docs, I think.

dberenbaum replied:
> Do you mean linking from here?

No, just thinking out loud that this idea would be good to incorporate into the blog post.

jorgeorpinel commented:
@shcheklein @iesahin @dberenbaum I updated the PR description to reflect that it's most important aspect is now introducing `exp init`. Moved some tasks to possible follow-ups in case we want to merge this by the release date and extract those discussions to a separate issue.

dberenbaum commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,75 @@
+
+[run them]: /doc/user-guide/experiment-management/running-experiments
+
+## Properties
+
+DVC Experiments will have an auto-generated name like `exp-bfe64` by default. A
+custom name can be given instead (using the `--name`/`-n` option of
+`dvc exp run`). These names can be used to reference experiments in other
+`dvc exp` subcommands.
+
+All experiments created by DVC will be associated to the latest commit (Git
+`HEAD`) at the time that they were run. This is called the experiment's
+_baseline_.
```
Is this section needed? Naming details doesn't seem that important, and the baseline is already explained above.

jorgeorpinel replied:
Hmm good points. I moved the naming paragraph to the hidden section for now (having it here instead of in the Running Exps guide was one of the original goals of the overview page). Removed the duplicate "baseline" explanation. Thanks

dberenbaum commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,75 @@
+
+</details>
+
+[run them]: /doc/user-guide/experiment-management/running-experiments
```
There is a workflow section with this link below. Is it also needed here?

jorgeorpinel replied:
Fixed.

dberenbaum commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,75 @@
+
+`dvc exp init` lets you onboard any existing data science project to use DVC
+Experiments without having to worry bootstrapping DVC manually. It will prompt
+you wth a few simple questions and create a basic `dvc.yaml` file, as well as
```
```suggestion
you with a few simple questions and create a basic `dvc.yaml` file, as well as
```

jorgeorpinel replied:
Fixed.

dberenbaum commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,75 @@
+means that minimal formalities are required.
+
+`dvc exp init` lets you onboard any existing data science project to use DVC
+Experiments without having to worry bootstrapping DVC manually. It will prompt
```
```suggestion
Experiments without having to worry bootstrapping DVC manually. `dvc exp init -i` will prompt
```

Might want to somehow specify to use `--interactive/-i` if mentioning prompts.

jorgeorpinel replied:
Forgot about `-i` ⌛

dberenbaum replied:
Actually, we probably want to focus primarily on `dvc exp init -i` throughout this section. That's really the beginner-friendly way to get started.

jorgeorpinel replied:
@dberenbaum I think both a simple `command` and `-i` are equally beginner-friendly. Mentioned both now.

dberenbaum commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,75 @@
+`dvc exp init` lets you onboard any existing data science project to use DVC
+Experiments without having to worry bootstrapping DVC manually. It will prompt
+you wth a few simple questions and create a basic `dvc.yaml` file, as well as
+other <abbr>metafiles</abbr> with sane default values. You can review these
```
I don't think it actually creates any other files. It also might not even create `dvc.yaml` if one already exists, but not sure that's an important distinction here.

jorgeorpinel replied:
> I don't think it actually creates any other files

Hmmm so how does it work? I guess at this point I need to go and actually use it again 😅 

It just grabs defaults from a new config folder right? Checking ⌛

dberenbaum replied:
It uses either the paths provided in the prompts/arguments, the config file, or the defaults (in that priority) to generate a stage in `dvc.yaml`.

jorgeorpinel replied:
Got it. Updated, PTAL

jorgeorpinel replied:
> the config file

p.s. I didn't mention config files to avoid complicating here. In fact this isn't very well explained in the ref. either -- I'll look into that.

dberenbaum commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,75 @@
+means that minimal formalities are required.
+
+`dvc exp init` lets you onboard any existing data science project to use DVC
+Experiments without having to worry bootstrapping DVC manually. It will prompt
```
```suggestion
Experiments without having to worry about bootstrapping DVC manually. It will prompt
```

dberenbaum commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,75 @@
+Experiments without having to worry bootstrapping DVC manually. It will prompt
+you wth a few simple questions and create a basic `dvc.yaml` file, as well as
+other <abbr>metafiles</abbr> with sane default values. You can review these
+files and commit them to Git to begin using DVC Experiments quickly.
```
You shouldn't even need to commit to Git.

jorgeorpinel replied:
> shouldn't even need to commit

Yeah but... You usually want to? Or I guess it's up to the user... ⌛

dberenbaum replied:
True. Not a big deal either way.

jorgeorpinel replied:
I put this in parenthesis now.

dberenbaum commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,75 @@
+
+`dvc exp init` lets you onboard any existing data science project to use DVC
+Experiments without having to worry bootstrapping DVC manually. It will prompt
+you wth a few simple questions and create a basic `dvc.yaml` file, as well as
```
The rest of the paragraph after "create a basic `dvc.yaml` file` might be unnecessary. `dvc.yaml` also probably is unnecessary to introduce at all. You could probably combine this sentenc with the first sentence of the following paragraph, like:

> `dvc exp init -i` will prompt you with a few simple questions and create a basic pipeline (even if it has a single stage that represents your entire process).

jorgeorpinel replied:
Gone.

iesahin commented on content/docs/command-reference/exp/apply.md:
```diff
@@ -15,21 +15,25 @@ positional arguments:
+<abbr>parameters</abbr>, <abbr>metrics</abbr>, etc.) needed to reflect the
+experiment conditions and results in the workspace.
+
+⚠️ Current changes to the workspace are preserved except if they conflict with
```
I think this word of caution is phrased too long. Earlier version is more to the point.

jorgeorpinel replied:
Yeah but we wanted to add that first part. I moved it to the paragraph above to keep the warning short.

iesahin commented on content/docs/command-reference/exp/apply.md:
```diff
@@ -15,21 +15,25 @@ positional arguments:
-[persistent](/doc/user-guide/experiment-management#persistent-experiments)).
+or `dvc exp diff`, and before committing it to Git (making it [persistent].
+
+Note that the history of [checkpoints] found in the `experiment` is **not**
```
It might be better to have a ⚠️ here?

jorgeorpinel replied:
Idk. This mention of checkpoints is already kind of out of nowhere, especially now that checkpoints info is not in the cmd ref. (updated link to guide). I changed it a bit and made it a note for now.

iesahin commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,73 @@
+means that minimal formalities are required.
+
+`dvc exp init` lets you onboard any existing data science project to use DVC
+Experiments without having to worry about bootstrapping DVC manually. This
```
The user still needs to run `dvc init` before `dvc exp init` though.

jorgeorpinel replied:
Good point! I added that note to the `exp init` ref instead of complicating this text.

iesahin reviewed (commented):
Some minor points.

dberenbaum commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,71 @@
+Experiments preserve a connection to the latest commit in the current branch
+(Git `HEAD`) as their parent or _baseline_, but do not form part of the regular
+Git tree or workflow (unless you make them [persistent]). This prevents
+polluting Git namespaces and bloating the repo unnecessarily.
```
Another option might be to reference that it's explained in more detail below.

jorgeorpinel replied:
> this paragraph feels a little abstract

Updated a little (3af2f9ac)

> Do we want to avoid terms like ephemeral/lightweight branches

Yes, those terms would also require definition and they're an implementation detail anyway. We're trying to stick to what matters most from the user perspective.

> reference that it's explained in more detail below

The title of the hidden section which is right after this paragraph should already serve as a reference or call-out for more info, I think.

dberenbaum commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,71 @@
+DVC Experiments build on basic semantics of <abbr>DVC projects</abbr>. This
+means that minimal formalities are required.
+
+`dvc exp init` lets you quickly onboard an existing data science project to use
+DVC Experiments, without having to worry about bootstrapping DVC manually. You
+can either supply a `command` to execute your experiments or use the
+`--interactive` flag (`-i`) to be prompted for that and other optional
+customizations.
```
I think pushing `-i` hard here is worthwhile since it is by far the easiest way I have found to use the command, especially for beginners. Do we even need to mention `supply a command to execute your experiments`?

dberenbaum commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,71 @@
+`--interactive` flag (`-i`) to be prompted for that and other optional
+customizations.
+
+This creates a simple `dvc.yaml` file for you. It uses sane default locations
```
Should we explain what `dvc.yaml` is at all or at least mention that it creates a pipeline stage?

dberenbaum commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,71 @@
+
+## Initialize DVC Experiments on any project
+
+DVC Experiments build on basic semantics of <abbr>DVC projects</abbr>. This
+means that minimal formalities are required.
```
> This means that minimal formalities are required.

I'm not sure what this is intended to mean. I can read it to mean pretty much opposite things (either DVC doesn't have a lot of formalities or that DVC requires some formalities), so I find it confusing. Is it needed?

jorgeorpinel replied:
Hmmmm how about "To use DVC Experiments you need a <abbr>DVC project</abbr> with a minimal
structure and configuration" ?

dberenbaum commented on content/docs/user-guide/experiment-management/experiments-overview.md:
```diff
@@ -0,0 +1,71 @@
+`--interactive` flag (`-i`) to be prompted for that and other optional
+customizations.
+
+This creates a simple `dvc.yaml` file for you. It uses sane default locations
```
Can we jump right into customizable locations instead of first focusing on sane defaults (I would hope/assume it has sane defaults!).

jorgeorpinel replied:
Well, the latest version is pretty different, want to take a look? Currently I have to mention the defaults before how to customize them.

Also I added a hypothetical link to a change I'm making in https://github.com/iterative/dvc.org/pull/3071 that specifies the default project structure (borrowed from the blog post).

dberenbaum reviewed (commented):
A few more minor comments 🙏 . If it's vital to get it merge now, let me know and I can approve, but I think one more iteration might be useful.

jorgeorpinel commented:
@shcheklein several possible follow-ups here but do you think this is mergeable? Thanks

jorgeorpinel commented:
UPDATE: I removed a bunch of unrelated changes that somehow got to this PR (probably nested PR), moving to https://github.com/iterative/dvc.org/pull/3080

jorgeorpinel commented:
I know there are still questions around this but they're larger than just about the index and overview page so I'm merging this much. Should be followed upon in https://github.com/iterative/dvc.org/issues/2911. Thanks
